controlplane/internal/controller/agentcontrolplane_controller.go

```diff
@@ -199,6 +216,70 @@ func (r *AgentControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	return ctrl.Result{}, rerr
 }
 
+// Ensures dependencies are deleted before allowing the AgentControlPlane to be deleted
+// Deletes the ClusterDeployment (which deletes the AgentClusterInstall)
+// Machines, InfraMachines, and AgentBootstrapConfigs get auto-deleted when the ACP has a deletion timestamp - this deprovisions the BMH automatically
+// TODO: should we handle watching until all machines & agentbootstrapconfigs are deleted too?
+func (r *AgentControlPlaneReconciler) handleDeletion(ctx context.Context, acp *controlplanev1beta1.AgentControlPlane) (ctrl.Result, error) {
```

Comment:
This methods is quite long, can we extract `deleteClusterDeployment(...)` and `deleteInfraEnv(...)`?

Reply:
Btw, not sure we should handle deleting infraenv from this controller, as the infraenv is being created by AgentBootstrap
Wouldn't the approach of letting cluster deployment cascade deletion handle it work?

Reply:
I tested it out, and was wrong - the clusterDeployment doesn't actually cascade delete the infraenv :(

Reply:
Also remember we talked about the AgentBootstrapConfig is not a one-to-one mapping to infraEnv so we can't just delete an infraEnv everytime we delete an AgentBootstrapConfig because of how it's currently designed. It would need to be by Control Plane or by Machine Deployment because those have a 1:1 mapping to infra envs

Reply:
Ok, if we can't leverage cascade deleting, I'd propose the following:
* we set the finalizer on infraenv from agentbootstrapconfig
* we can index `bootstrapConfig.Status.InfraEnvRef` field
* on deletion, we can check if there is any other resource referencing such infraEnv, if not we can handle deletion

The advantages I can see of the above approach is the following:
* we keep the code related to creating and deleting infraenv in the same package/controller
* we don't need extra RBAC permissions in other controllers to handle infraenv

On the other hand, it'll be slightly more complicated than just sharing logic between the controllers as we are doing in this PR.

WDYT?

Reply:
Yeah, if you agree I would delegate InfraEnv handling to the bootstrap provider

Reply:
Yep! Thanks Riccardo!